src/main/java/com/mongodb/spark/sql/connector/connection/MongoClientCache.java

```diff
@@ -139,36 +138,37 @@ private void assertIsAvailable() {
     private final MongoClientFactory mongoClientFactory;
     private final long keepAliveNanos;
     private final MongoClient mongoClient;
-    private long releaseTime;
+    private boolean acquired;
+    private long releasedNanos;
     private int referenceCount;
 
     private MongoClientCacheItem(
         final MongoClientFactory mongoClientFactory, final long keepAliveNanos) {
       this.mongoClientFactory = mongoClientFactory;
       this.mongoClient = mongoClientFactory.create();
       this.keepAliveNanos = keepAliveNanos;
-      this.releaseTime = Long.MAX_VALUE;
+      this.releasedNanos = System.nanoTime();
       this.referenceCount = 0;
     }
 
     private MongoClient acquire() {
       referenceCount += 1;
-      releaseTime = Long.MAX_VALUE;
+      acquired = true;
       return mongoClient;
     }
 
     private void release() {
+      Assertions.ensureState(
+          () -> referenceCount > 0, "MongoClient reference count cannot be below zero");
+      releasedNanos = System.nanoTime();
       referenceCount -= 1;
-      if (referenceCount == 0) {
-        releaseTime = System.nanoTime() + keepAliveNanos;
-      }
     }
 
-    private boolean shouldBeRemoved(final long currentTime) {
-      if (referenceCount == 0 && releaseTime < currentTime) {
+    private boolean shouldBeRemoved(final long currentNanos) {
+      if (acquired && referenceCount == 0 && currentNanos - releasedNanos > keepAliveNanos) {
```

Comment:
While it's true that the `acquired` check would have eliminated the race condition that would have allowed the method `MongoClientCache.acquire` to return a closed ` MongoClient`, the `MongoClientCache` as it is currently written does not have this race condition anyway because executions of the methods `MongoClientCache.acquire` and `MongoClientCache.checkClientCache` are sequential (both methods are `synchronized`).

The `acquired` field serves no purposes even if we say that `MongoClientCacheItem` must not rely on the knowledge of how it is used (which is a valid design decision, but the alternative is also valid: it may rely on that knowledge as it's a `private` static class nested in `MongoClientCache`), because the `MongoClientCacheItem` is not thread-safe, and, thus, already relies on the correct synchronization provided by `MongoClientCache`.